src/frequenz/sdk/microgrid/_data_pipeline.py

```diff
@@ -0,0 +1,207 @@
+    channel: Broadcast["ComponentMetricRequest"]
+
+
+class _DataPipeline:
+    """Create, connect and own instances of data pipeline components.
+
+    Provides SDK users direct access to higher level components of the data pipeline,
+    eliminating the boiler plate code required to setup the DataSourcingActor and the
+    ResamplingActor.
+    """
+
+    def __init__(
+        self,
+        resampler_config: ResamplerConfig,
+    ) -> None:
+        """Create a `DataPipeline` instance.
+
+        Args:
+            resampler_config: Config to pass on to the resampler.
+        """
+        from ..actor import ChannelRegistry
+
+        self._resampler_config = resampler_config
+
+        self._channel_registry = ChannelRegistry(name="Data Pipeline Registry")
+
+        self._data_sourcing_actor: _ActorInfo | None = None
+        self._resampling_actor: _ActorInfo | None = None
+        self._logical_meter: "LogicalMeter" | None = None
+        self._ev_charger_pools: dict[frozenset[int], "EVChargerPool"] = {}
+
+    def logical_meter(self) -> LogicalMeter:
+        """Return the logical meter instance.
+
+        If a LogicalMeter instance doesn't exist, a new one is created and returned.
+
+        Returns:
+            A logical meter instance.
+        """
+        from ..microgrid import connection_manager
+        from ..timeseries.logical_meter import LogicalMeter
+
+        if self._logical_meter is None:
+            self._logical_meter = LogicalMeter(
+                channel_registry=self._channel_registry,
+                resampler_subscription_sender=self._resampling_request_sender(),
+                component_graph=connection_manager.get().component_graph,
+            )
+        return self._logical_meter
+
+    def ev_charger_pool(
+        self,
+        ev_charger_ids: set[int] | None = None,
+    ) -> EVChargerPool:
+        """Return the corresponding EVChargerPool instance for the given ids.
+
+        If an EVChargerPool instance for the given ids doesn't exist, a new one is
+        created and returned.
+
+        Args:
+            ev_charger_ids: Optional set of IDs of EV Charger to be managed by the
+                EVChargerPool.
+
+        Returns:
+            An EVChargerPool instance.
+        """
+        from ..timeseries.ev_charger_pool import EVChargerPool
+
+        # We use frozenset to make a hashable key from the input set.
+        key: frozenset[int] = frozenset()
+        if ev_charger_ids is not None:
+            key = frozenset(ev_charger_ids)
+
+        if key not in self._ev_charger_pools:
+            self._ev_charger_pools[key] = EVChargerPool(
+                channel_registry=self._channel_registry,
+                resampler_subscription_sender=self._resampling_request_sender(),
+                component_ids=ev_charger_ids,
+            )
+        return self._ev_charger_pools[key]
+
+    def _data_sourcing_request_sender(self) -> Sender[ComponentMetricRequest]:
```

Comment:
documentation is missing for this function

Reply:
done, for the other private functions as well.